File: oss-center/src/main/java/com/jmsoftware/maf/osscenter/read/service/ReadResourceService.java
package com.jmsoftware.maf.osscenter.read.service;

import com.jmsoftware.maf.osscenter.read.entity.SerializableStatObjectResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.lang.Nullable;
import org.springframework.util.unit.DataSize;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;

import javax.validation.constraints.NotBlank;

/**
 * <h1>ReadResourceService</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (鍾俊), email: johnnysviva@outlook.com, 6/20/21 5:17 PM
 **/
@Validated
public interface ReadResourceService {
    String BUCKET_OBJECT_NAME_REGEX = "^.+/.+$";
    DataSize SMALL_CHUNK_SIZE = DataSize.ofMegabytes(1);
    DataSize MEDIUM_CHUNK_SIZE = DataSize.ofMegabytes(4);
    DataSize LARGE_CHUNK_SIZE = DataSize.ofMegabytes(8);

    /**
     * Async stream single resource response entity.
     *
     * @param bucket the bucket
     * @param object the object
     * @param range  the range
     * @return the response entity
     */
    ResponseEntity<StreamingResponseBody> asyncStreamSingleResource(@NotBlank String bucket, @NotBlank String object,
                                                                    @Nullable String range);

    /**
     * Async download single resource response entity.
     *
     * @param bucket the bucket
     * @param object the object
     * @return the response entity
     */
    ResponseEntity<StreamingResponseBody> asyncDownloadSingleResource(@NotBlank String bucket, @NotBlank String object);

    /**
     * Gets resource information.
     *
     * @param bucket the bucket
     * @param object the object
     * @return the resource detail
     */
    SerializableStatObjectResponse stateObject(@NotBlank String bucket, @NotBlank String object);
}

File: oss-center/oss-center-biz/src/main/kotlin/com/jmsoftware/maf/osscenter/service/ReadResourceService.kt
package com.jmsoftware.maf.osscenter.service

import com.jmsoftware.maf.osscenter.response.SerializableStatObjectResponse
import jakarta.validation.constraints.NotBlank
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody

/**
 * # ReadResourceService
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/16/22 11:48 PM
 */
@Validated
interface ReadResourceService {
    /**
     * Get single resource, especially pictures, or other small size file.
     *
     * @param bucket the bucket
     * @param object the object
     * @return the response entity
     */
    fun asyncGetSingleResource(
        @NotBlank bucket: String,
        @NotBlank `object`: String
    ): ResponseEntity<StreamingResponseBody>

    /**
     * Stream single resource, more efficiency if the resource is large file, which will be streamed by range.
     * Responsive streaming
     *
     * @param bucket the bucket
     * @param object the object
     * @param range  the range
     * @return the response entity
     */
    fun asyncStreamSingleResource(
        @NotBlank bucket: String,
        @NotBlank `object`: String,
        range: String?
    ): ResponseEntity<StreamingResponseBody>

    /**
     * Async download single resource response entity.
     *
     * @param bucket the bucket
     * @param object the object
     * @return the response entity
     */
    fun asyncDownloadSingleResource(
        @NotBlank bucket: String,
        @NotBlank `object`: String
    ): ResponseEntity<StreamingResponseBody>

    /**
     * Gets resource information.
     *
     * @param bucket the bucket
     * @param object the object
     * @return the resource detail
     */
    fun stateObject(
        @NotBlank bucket: String,
        @NotBlank `object`: String
    ): SerializableStatObjectResponse
}
